The ViewModels lists composition dance (#113)
mauroservienti/milestone.github.io: _posts/2019-03-21-the-viewmodels-lists-composition-dance.md

typora-root-url: ..
layout: post
header_image: /img/posts/the-viewmodels-lists-composition-dance/header.jpg
title: "The ViewModels Lists Composition dance"
author: Mauro Servienti
synopsis: "Composing a list in an efficient way is complex. Multiple components need to interact with each other to produce the final result. It's not that different from a chorus, where each dancer performs the studied choreography so that the end result is an amazing ballet."
tags:
- SOA
- Services ViewModel Composition
---

We've recently had a [look at some pseudo code showing how Single Item Composition could work](https://milestone.topics.it/2019/03/06/viewmodel-composition-show-me-the-code.html). At the end of that article the composition flow was summarized using the following diagram:

![composition engine information flow](https://milestone.topics.it/img/posts/viewmodel-composition-show-me-the-code/1551868103502.png){:class="img-responsive"}

The diagram is a bird-eye view of the composition flow, and misses details about lists composition whose [darkness we've already investigated](https://milestone.topics.it/2019/02/28/into-the-darkness-of-viewmodel-lists-composition.html).

## Let's dance

> In the mentioned article, _Marketing_ is identified as the *Product* concept owner, and as such _Marketing_ is the one responsible to return the **list** of *Products*:

Author: ```suggestion
In the mentioned articles, _Marketing_ is identified as the one owning the *Product* concept, and as such is responsible to return the **list** of *Products*:
```

```csharp
class MarketingProductsListGetHandler : IHandleRequests
{
   public bool Matches(RouteData routeData, string httpVerb, HttpRequest request)
   {
      var controller = (string)routeData.Values["controller"];
      var action = (string)routeData.Values["action"];

      return HttpMethods.IsGet(httpVerb)
         && controller.ToLowerInvariant() == "products"
         && action.ToLowerInvariant() == "index"
         && !routeData.Values.ContainsKey("id");
   }
}
```

First things, first. The `MarketingProductsListGetHandler` is interested in all requests that are for the `products` controller and don't contain an `id`. In essence URLs like `/products`.

```csharp
class MarketingProductsListGetHandler : IHandleRequests
{
   public bool Matches(RouteData routeData, string httpVerb, HttpRequest request)
   {
      /* omitted for clarity */
   }
   
   public async Task Handle(dynamic vm, RouteData routeData, HttpRequest request)
   {
      var url = $"http://marketing.backend.local/api/available/products";
      var response = await new HttpClient().GetAsync(url);

      var availableProducts = await response.Content.As<int[]>();
   }
}
```

> Whenever a matching request is intercepted, _Marketing_ retrieves the list of available products, in the above snippet just IDs are returned.

Author: ```suggestion
Whenever a matching request is intercepted, _Marketing_ retrieves the list of available products. Products that in the above snippet are just IDs.
```

   {
      var url = $"http://marketing.backend.local/api/available/products";
      var response = await new HttpClient().GetAsync(url);

      var availableProducts = await response.Content.As<int[]>();
      
      var availableProductsViewModel = MapToDictionary(availableProducts);
      await vm.RaiseEvent(new AvailableProductsLoaded()
      {
         AvailableProductsViewModel = availableProductsViewModel
      });
      
      vm.AvailableProducts = availableProductsViewModel.Values.ToList();
   }
}
```

The first thing is to prepare a data structure suitable for event handlers that need to augment each _Product_. `MapToDictionary` is as simple as:

```csharp
IDictionary<int, dynamic> MapToDictionary(IEnumerable<int> availableProducts)
{
   var availableProductsViewModel = new Dictionary<int, dynamic>();

   foreach (var id in availableProducts)
   {
      dynamic vm = new ExpandoObject();
      vm.Id = id;

      availableProductsViewModel[id] = vm;
   }

   return availableProductsViewModel;
}
```

Once that's ready the `AvailableProductsLoaded` event is raised, and finally the composed ViewModels are appended to the response ViewModel (`vm`).

> > The way the event is raised and dispatched to handlers is still obscure. As obscure is the whole composition infrastructure. Don't desperate there will be a dedicated article.

Author: ```suggestion
> The way the event is raised and dispatched to handlers is still obscure. As obscure is the whole composition infrastructure. Don't desperate, there will be a dedicated article.
```

## False notes

> The `IHandleRequests` interfaces is clearly insufficient. Semantically it's not designed to handle events.

Author: ```suggestion
The `IHandleRequests` interfaces is clearly insufficient. Semantically, it's not designed to handle events.
```

```csharp
interface ISubscribeToCompositionEvents
{
   void Subscribe(ISubscriptionStorage subscriptionStorage, RouteData routeData, HttpRequest request);
    }
```

`ISubscribeToCompositionEvents` expresses the intent very well. Components implementing it are interested in subscribing to composition events. For example, _Sales_ could do something like:

```csharp
class AvailableProductsLoadedSubscriber : ISubscribeToCompositionEvents
{
   public void Subscribe(ISubscriptionStorage subscriptionStorage, RouteData routeData, HttpRequest request)
   {
      subscriptionStorage.Subscribe<AvailableProductsLoaded>(async (requestId, pageViewModel, @event, rd, req) =>
      {
         //whatever sales need to do
      });
   }
}
```

> At runtime the composition infrastructure will invoke all components that implement the `ISubscribeToCompositionEvents` interface giving them the opportunity to subscribe to composition events. In e above snippet _Sales_ is subscribing to the `AvailableProductsLoaded` event that _Marketing_ raises whenever a list of available products is requested. When _Sales_ handles the event, additional products information are retrieved and composed into the list:

Author: ```suggestion
At runtime the composition infrastructure will invoke all components implementing the `ISubscribeToCompositionEvents` interface giving them the opportunity to subscribe to composition events. In the above snippet _Sales_ is subscribing to the `AvailableProductsLoaded` event that _Marketing_ raises whenever a list of available products is requested.

When _Sales_ handles the event, additional products information are retrieved and composed into the list:
```
